### ISSUE TITLE
Add JSON schema support

### ISSUE THREAD

**Issue by sdeleuze** [spring-projects/spring-hateoas#444]
It would be super useful for Spring Hateoas to support additional documentation metadata like being able to describe JSON data with [JSON schema](http://json-schema.org/).

See also https://github.com/spring-projects/spring-hateoas/issues/16 for related discussion.

**Comment by olivergierke** [comment]
What would that support look like? Spring HATEOAS doesn't know anything about the structure of the application level content. All it currently cares about is hypermedia controls and how to create them. Note that there is JSON Schema support in Spring Data REST, where I think it makes a lot of sense, as — through the Spring Data `PersistentEntity` API — that library has information about the parts of the representation that resemble domain types, which basically means that there's something to actually create a schema for.

**Comment by sdeleuze** [comment]
My use case is really about allowing JSON schema support in any kind of RESTful Spring Boot + `@RestController`applications, not only Spring Data REST based ones.

I think such feature would help to provide an alternative with Spring HATEOAS + [Spring REST Docs](http://projects.spring.io/spring-restdocs/) (cc @wilkinsona) to [OpenAPI](https://openapis.org/) (very Swagger oriented, and not HATEOAS compliant at all) or [RAML](http://raml.org/) that are documenting a wide range of informations (Spring REST docs provides human readable specs but not machine readable ones), and is complementary to [this PR about providing title and type properties](https://github.com/spring-projects/spring-hateoas/pull/79).

Such additional metadata could be used for various purpose:
 - Automated testing of RESTful APIs
 - Developer documentation for free
 - Expose more information for libraries like [Orbit.js](https://github.com/orbitjs/orbit.js/)

As mentioned in this [blog post](https://opencredo.com/hal-hypermedia-api-spring-hateoas/#comment-115) or [implemented by Spring REST Docs](http://docs.spring.io/spring-restdocs/docs/current/reference/html5/#documenting-your-api-constraints), it could maybe use by default the field names, types and Bean Validation annotations, while providing some extension points to add/customize contraints.

Any thoughts?

**Comment by gregturn** [comment]
The only avenue I see to implement something would be providing an API to build a JSON Schema complaint document. But it would be to the library user to actually use it. 

I don't know how easy or hard it is to build that given Spring HATEOAS isn't setup for building metadata documents.

**Comment by sdeleuze** [comment]
Indeed, that's maybe outside of Spring HATEOAS scope. Maybe supporting JSON schema would be better at Spring REST docs level ([see this issue](https://github.com/spring-projects/spring-restdocs/issues/43)) or at [RAML](http://raml.org/) level for those who seeks machine readable specs ...

See [this discussion about HATEOAS support in RAML forums](http://forums.raml.org/t/raml-versus-hateoas/44).

**Comment by gregturn** [comment]
@odrotbohm can we simply move the JSON Schema builders from Spring Data REST to Spring HATEOAS?

SDR can continue to use them. Spring HATEOAS can now offer them to anyone _else_ that wants to manually assemble such a docu.
